src/libraries/System.Net.Http/src/System/Net/Http/Metrics/HttpMetricsEnrichmentContext.cs

```diff
@@ -0,0 +1,162 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Net.Http.Metrics;
+
+public class HttpMetricsEnrichmentContext
+{
+    private HttpRequestMessage? _request;
+    private HttpResponseMessage? _response;
+    private Exception? _exception;
+    private TagCollection _tags;
+
+    public HttpMetricsEnrichmentContext()
+    {
+        _tags = new TagCollection();
+    }
+
+    public HttpRequestMessage Request
+    {
+        get
+        {
+            EnsureRunningCallback();
```

Comment:
I think you get similar level of protection by clearing the various fields after the callback is complete rather than adding explicit checks.